vortex-scan/src/lib.rs

```diff
@@ -330,5 +314,131 @@ fn to_field_mask(field: FieldName) -> FieldMask {
     FieldMask::Prefix(FieldPath::from(Field::Name(field)))
 }
 
+/// A projected subset (by indices, range, and filter) of rows from a Vortex data source.
+///
+/// The method of this struct enable, possibly concurrent, scanning of multiple row ranges of this
+/// data source.
+///
+/// See also: [ScanBuilder].
+pub struct RepeatedScan<A: 'static + Send> {
+    layout_reader: LayoutReaderRef,
+    projection: ExprRef,
+    filter: Option<ExprRef>,
+    /// Optionally read a subset of the rows in the file.
+    row_range: Option<Range<u64>>,
+    /// The selection mask to apply to the selected row range.
+    selection: Selection,
+    /// The natural splits of the file.
+    splits: Vec<Range<u64>>,
+    /// The number of splits to make progress on concurrently **per-thread**.
+    concurrency: usize,
+    /// Function to apply to each [`ArrayRef`] within the spawned split tasks.
+    map_fn: Arc<dyn Fn(ArrayRef) -> VortexResult<A> + Send + Sync>,
+    /// Maximal number of rows to read (after filtering)
+    limit: Option<usize>,
+    /// The dtype of the projected arrays.
+    dtype: DType,
+}
+
+impl<A: 'static + Send> RepeatedScan<A> {
+    pub fn execute(
+        &self,
+        row_range: Option<Range<u64>>,
+    ) -> VortexResult<Vec<BoxFuture<'static, VortexResult<Option<A>>>>> {
+        let row_range = intersect_ranges(self.row_range.as_ref(), row_range);
+
+        let ctx = Arc::new(TaskContext {
+            row_range,
+            selection: self.selection.clone(),
+            filter: self.filter.clone().map(|f| Arc::new(FilterExpr::new(f))),
+            reader: self.layout_reader.clone(),
+            projection: self.projection.clone(),
+            mapper: self.map_fn.clone(),
+        });
+
+        // Create a task that executes the full scan pipeline for each split.
+        let mut limit = self.limit;
+        let split_tasks = self
+            .splits
+            .iter()
+            .filter_map(|split_range| {
+                if limit.is_some_and(|l| l == 0) {
+                    None
+                } else {
+                    Some(split_exec(ctx.clone(), split_range.clone(), limit.as_mut()))
+                }
+            })
+            .try_collect()?;
+
+        Ok(split_tasks)
+    }
+
+    #[cfg(feature = "tokio")]
+    pub fn execute_tokio_stream(
```

Comment:
calling this into when it takes &self, instead of self is wrong (JetBrains IDEs warn you about it)